Create a basic enum validation source generator

davidwengier opened the pull request:
You convinced me @RussKie I'll port my sample over here, and see if I can make it a little better :)

* [x] Add to the solution
* [x] Handle flags enums
* [x] Use a SyntaxReciever for perf
* [x] Tests

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/4043)

RussKie commented on src/System.Windows.Forms.SourceGenerators/System.Windows.Forms.SourceGenerators.csproj:
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Just to clarify, adding this project does not affect our shippable state (i.e. we don't need to package and bundle it), and it is compile time only, right?

davidwengier replied:
Correct

RussKie reviewed (commented):
🚀

davidwengier commented:
Well it built! Not bad for v1 :P

RussKie commented:
> Well it built! Not bad for v1 :P

Hmm... did our tests worked? 🤔 
https://github.com/dotnet/winforms/blob/fc75f9c961b83d96e399afe4f5384798f59c9b45/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ButtonTests.cs#L275-L611

J/k well done!

davidwengier commented:
Worth noting: Having a source generator means contributors using Visual Studio need to be on at least 16.8 Preview 3.1. That might not be in preview anymore by the time this is finished, but thought I'd mention it. VS Code and command line builds should work fine.

davidwengier commented:
Wow, there are even tests for enum validity already?! I didn't even check yet!

JeremyKuhne commented:
@davidwengier How does the generated code integrate with the debugging experience? Both in the development experience in VS and trying to debug dumps.

davidwengier commented:
In VS, debugging (eg, step into etc.) and Go To Definition work the same as any other code that is in the consuming project, including seeing comments, indentation etc. ie, its better than Metadata as Source (though still read only). The generated code even live-updates as you code in VS. In general though I would consider the features in VS to be at beta level, and there are more improvements coming.

For dump debugging I'm not sure actually. PDBs and debug info would be unaffected, but source files don't exist (by default) so not sure what story is around symbol servers etc. @chsienki hopefully can answer that one.

RussKie commented on src/System.Windows.Forms.SourceGenerators/EnumValidationGenerator.cs:
```diff
@@ -0,0 +1,178 @@
+    {
+        private const string EnumValidatorStub = @"
+// <auto-generated />
+namespace EnumValidation
```
To make it more obvious where this code came from and how to deal with it, perhaps call it:
```suggestion
namespace SourceGenerated
```

The same applies below

RussKie replied:
This is still outstanding.

davidwengier replied:
I'm gonna do this, but I don't think it will be very nice to use in practice, and won't be surprised if its changed back later. Why does a developer need to care that the enum validation is generated? I actually think the best namespace for the validator is probably `System.Windows.Forms` so its even easier to find.

RussKie commented on src/System.Windows.Forms.SourceGenerators/EnumValidationGenerator.cs:
```diff
@@ -0,0 +1,178 @@
+                throw new InvalidOperationException("We were given the wrong syntax receiver.");
+            }
+
+            var enumsToValidate = GetEnumValidationInfo(context.Compilation, syntaxReceiver.ArgumentsToValidate);
```
Instead of `var` please use actual types, makes it easier to understand the code.

davidwengier replied:
Why isn't there an .editorconfig rule preventing me from doing the wrong thing? 😛

RussKie replied:
The general limitations of 24-hour day, whoever made the decision to stop at 24 and not 80 or 100 made a bad call :D


Besides enabling this rule would involve a bit of clean up. You're welcome to do this separately, haha

RussKie commented on src/System.Windows.Forms.SourceGenerators/EnumValidationGenerator.cs:
```diff
@@ -0,0 +1,178 @@
+
+            foreach (var info in infos)
+            {
+                string indent = "        ";
```
Hoist it outside the loop and turn into a const?

RussKie commented on src/System.Windows.Forms.SourceGenerators/EnumValidationGenerator.cs:
```diff
@@ -0,0 +1,178 @@
+            {
+                total |= element.Value;
+            }
+            sb.AppendLine($"{indent}if ((intValue & {total}) == intValue) return;");
```
Is this the same as `total.HasFlags(intValue)`?

davidwengier replied:
This is the same as the method it replaces, [`ClientUtils.IsEnumValid_Masked`](https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/ClientUtils.cs#L66)

RussKie commented on src/System.Windows.Forms.SourceGenerators/EnumValidationGenerator.cs:
```diff
@@ -0,0 +1,178 @@
+            int? max = null;
+            foreach (var info in elements)
+            {
+                if (max == null || info.Value != max + 1)
```
```suggestion
                if (max is null || info.Value != max + 1)
```

RussKie commented on src/System.Windows.Forms.SourceGenerators/EnumValidationGenerator.cs:
```diff
@@ -0,0 +1,178 @@
+                    max = info.Value;
+                }
+            }
+            if (max == null)
```
```suggestion
            if (max is null)
```

RussKie commented on src/System.Windows.Forms.SourceGenerators/EnumValidationGenerator.cs:
```diff
@@ -0,0 +1,178 @@
+                var semanticModel = compilation.GetSemanticModel(argument.SyntaxTree);
+
+                var enumType = semanticModel.GetTypeInfo(argument).Type;
+                if (enumType == null || foundTypes.Contains(enumType))
```
```suggestion
                if (enumType is null || foundTypes.Contains(enumType))
```

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/ButtonBase.cs:
```diff
@@ -697,10 +695,7 @@ public TextImageRelation TextImageRelation
-                    throw new InvalidEnumArgumentException(nameof(value), (int)value, typeof(TextImageRelation));
-                }
+                EnumValidation.EnumValidator.Validate(value);
 
```
❗  The original validation logic was a little more complex than just check between min and max.
The underlying values for `TextImageRelation` are 2^n, so here we need a different (i.e. an additional) kind of validation.

davidwengier replied:
I don't think there is a problem here. The generated code for this overload looks like this:
![image](https://user-images.githubusercontent.com/754264/95155561-27660c00-07e0-11eb-90c4-53e7f5e721d7.png)

I am working on the assumption that the previous code was more complex because it was more efficient to be more complex, without resorting to reflection. Since the source generator doesn't need to do that, the complexity is not required.

There is a separate issue of validating the enum, if for some reason its not valid to have values that aren't 2^n, but that task is better handled by an analyzer or unit test IMO. The previous methods also validated the arguments that were passed to them, in DEBUG anyway, to validate that changes to the enum were reflected in the validation calls, and again that's not necessary as the source generator knows all.

I do plan to write tests to validate the validation, which should hopefully make the capabilities clearer.

TL,DR: "Marty, you're not thinking fourth dimensionally!"

RussKie replied:
This is.... ummm... brilliant! Thank you for explaining it.

weltkante replied:
Sorry if I'm missing something, just jumping into the discussion, but this doesn't look correct

> I don't think there is a problem here. The generated code for this overload looks like this:

This is a Flags-enum, the posted code doesn't properly validate values of a flags enum that are 2^n, am I missing something or why "is this not a problem" ?

> I am working on the assumption that the previous code was more complex because it was more efficient to be more complex

It (also) was more complex because it handled the case that flags-enums allow any _combination_ of enum values, not just a single value?

> if for some reason its not valid to have values that aren't 2^n, but that task is better handled by an analyzer or unit test IMO

You want to do runtime validation of arguments passed by the user, which can be computed or stored in variables and are not forced to be constant, that can't be done by either of what you suggest. Some flags enums have invalid combinations, if the previous code detected those, this detection either needs to be done manually now (split off the generated argument validation) or the generator needs to be told of those somehow.

davidwengier replied:
> This is a Flags-enum

No, [it's not](https://github.com/dotnet/winforms/blob/master/src/System.Windows.Forms/src/System/Windows/Forms/TextImageRelation.cs), but maybe that is a bug?
Flags enums are supported here so if the `[Flags]` attribute is added to the enum definition, the generator will adjust the validation accordingly. See [VisualStyleRenderer](https://github.com/dotnet/winforms/pull/4043/files/af81c84f58e3a555fce980d36be9b083e082fb26#diff-b20900b55d4e8f037d3f0c67a10dbe17L307-L314) and associated tests that validate that.

> You want to do runtime validation of arguments passed by the user, which can be computed or stored in variables and are not forced to be constant, that can't be done by either of what you suggest.

I'm not sure I follow. Can you point me to an example in the code? I am (slowly) enabling the new validation only in places where it applies. If it ends up that the `ClientTools.IsEnumValidXXX` have no callers then they can be removed, but if they're still needed, then they can be left in place. If there are scenarios where the validation can only be done at runtime, then there it shall remain. Alternatively it might be possible to add an additional overload that can help. Without a concrete example its hard to say.

davidwengier replied:
Worth noting, if this is a special type of enum where 2^n values are required, and it should be validated as a flags enum, but isn't a flags enum, then its possible to still use this generator, we just need to decorate the enum with some attribute that the generator can pick up and validate as though its `[Flags]`.

weltkante replied:
> No, it's not

my bad, was confused by the discussion talking about 2^n value space and the fact that the values are taken from a flags-enum

> I'm not sure I follow. Can you point me to an example in the code? I am (slowly) enabling the new validation only in places where it applies.

Never mind, I was talking about that some flag-enums define flags for which not all combinations are possible, so the flags enum isn't 2^n. I don't actually know if there are validation calls which check for this in the helper code (instead of having explicit checks), but if you are replacing the individual calls case-by-case anyways and can notice any special coding thats good enough.

Sorry for the noise then, I was reading the comments differently from what you both meant.

RussKie replied:
All good, I had to <kbd>F12</kbd> through the code to really see what underlying values were in this case.

It appears we do have a number of enums (by design) that proxy other enums that have flags-like values (whether those are flags or not is immaterial). And `IsEnumValid` checks that the first-order enums don't accept flags.

davidwengier replied:
>  some flag-enums define flags for which not all combinations are possible

This is a very good point, and definitely something to add a test case for.

> Sorry for the noise then

Absolutely not, all feedback is good. In particular, the validation I'm adding is thus far copying the existing validation that was done. There is a chance that existing validation is not comprehensive enough, or hasn't kept up, so its good to look at this with fresh eyes.

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs:
```diff
@@ -603,11 +604,7 @@ public DateTimePickerFormat Format
-                {
-                    throw new InvalidEnumArgumentException(nameof(value), (int)value, typeof(DateTimePickerFormat));
-                }
+                EnumValidator.Validate(value);
```
ditto

hughbe replied:
are we losing the `maxNumberOfBitsOn` validation?

davidwengier replied:
> are we losing the maxNumberOfBitsOn validation?

Yes, because I don't think it adds any value. We are now explicitly validating that the value is the same as one of the defined values in the enum. The only way `maxNumberOfBitsOn` would add value is for an enum defined with values that violate the validation (eg, having maxNumberOfBitsOn set to 1, but an enum with an element with a value of 3). I would think that if there was an enum defined like that, but validated with `maxNumberOfBitsOn` being 1, then thats just a honey pot trap for the user.

If there is an example you can point me to where `maxNumberOfBitsOn` adds value then I'll eat my hat (and make the appropriate code changes 😀)

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs:
```diff
@@ -1966,10 +1966,7 @@ public TextImageRelation TextImageRelation
-                {
-                    throw new InvalidEnumArgumentException(nameof(value), (int)value, typeof(TextImageRelation));
-                }
+                EnumValidation.EnumValidator.Validate(value);
```
ditto

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs:
```diff
@@ -192,19 +192,7 @@ public AutoCompleteSource AutoCompleteSource
-                {
-                    throw new InvalidEnumArgumentException(nameof(value), (int)value, typeof(AutoCompleteSource));
-                }
+                EnumValidation.EnumValidator.Validate(value);
```
Do you think we could have a different name that better communicates the validation?
E.g. something like:
```suggestion
                EnumValidation.EnumValidator.ValidateNotSequential(value);
```

davidwengier replied:
It's your codebase, but personally I think the power of the source generator is that it doesn't need this sort of distinction in order to work. You could create such a distinction, but you'd just end up with compile errors that amounted to: "You've done A. You meant to do B. I know you meant to do B, its the only option here, but I'm going to fail until you change the code to say 'B'". I think it would be annoying.

RussKie replied:
@JeremyKuhne what do you think?

hughbe replied:
> It's your codebase, but personally I think the power of the source generator is that it doesn't need this sort of distinction in order to work.

I agree with this point

JeremyKuhne replied:
> what do you think?

I don't think we should have different names. Above and beyond @davidwengier's point, perf notwithstanding we probably wouldn't create a generic `ValidateFlagsEnum<T>()` and `ValidateNonFlagsEnum<T>()`, nor would we want to go into significant detail `ValidateEnumValueIsDefined` or `ValidateFlagEnumValuesAreAllDefined`. It isn't terrible, but it doesn't add much value to the comprehension of the calling method and, in fact, may actually be more distracting than helpful.

RussKie reviewed (commented):
🚀

hughbe reviewed (commented):
This is awesome! Only thing I'd ask is to integrate the commit https://github.com/dotnet/winforms/pull/3204/commits/2994944597da3d4f5a54ea36c91bc4820afecce9 `Cleanup tests for valid/invalid enums` as that adds support for non sequential enums in `CommonTestHelper.GetEnumTypeTheoryDataInvalid` so we can have even more tests :)

hughbe commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DateTimePickerTests.cs:
```diff
@@ -201,6 +213,13 @@ public void DateTimePicker_SysTimeToDateTime_DoesnThrowException_If_SYSTEMTIME_I
         }
 
+        [WinFormsFact]
+        public void DateTimePicker_GetAutoSizeMode_Invoke_ReturnsExpected()
```
Nit: move this back to where it was - cleans up the diff

davidwengier replied:
I honestly don't remember moving this, and I can't work out why I would have ¯\\\_(ツ)_/¯

RussKie replied:
Please move it back.

JeremyKuhne commented on src/System.Windows.Forms.SourceGenerators/EnumValidationGenerator.cs:
```diff
@@ -0,0 +1,178 @@
+            {
+                string indent = "        ";
+
+                sb.AppendLine($"{indent}public static void Validate({info.EnumType} enumToValidate)");
```
Can we inject a `/// <summary>` line for the methods? I don't think we need full xml doc comments, but having the summary for Intellisense would be useful.

davidwengier commented:
Address most of the feedback, added some tests for the generation itself, and rolled out to way more places. I feel like this is close.

Open question for the readers (@RussKie @JeremyKuhne):

* Is there any value in the parameter name being optional? 
  * Originally I wanted the source generator to be smart and just know what the parameter name is, but that falls down once there is more than one call to Validate for the same enum type, so I've now added an optional parameter. Could just make it required?
* Can you think of some wording in the XML comment to tell a developer "If you don't see an overload that matches your enum here, just keep typing, trust me"?

A request: Would be interested in having someone pull this branch and use it for a bit. It's important that the presence of a source generator doesn't regress the editing experience in Visual Studio itself too, and we're keen on feedback for that too.

RussKie commented on src/System.Windows.Forms.SourceGenerators/tests/UnitTests/EnumValidationTests.cs:
```diff
@@ -0,0 +1,353 @@
+{
+    enum Names
+    {
+        David,
+        Igor,
+        Jeremy,
+        Hugh,
+        Tobias,
+        Olia,
+        Merrie
```
🤗 🎉

davidwengier commented on src/System.Windows.Forms/src/System/Windows/Forms/TableLayoutSettings.cs:
```diff
@@ -206,10 +203,7 @@ public TableLayoutPanelGrowStyle GrowStyle
                 //valid values are 0x0 to 0x2
-                if (!ClientUtils.IsEnumValid(value, (int)value, (int)TableLayoutPanelGrowStyle.FixedSize, (int)TableLayoutPanelGrowStyle.AddColumns))
-                {
-                    throw new ArgumentOutOfRangeException(nameof(value), value, string.Format(SR.InvalidArgument, nameof(GrowStyle), value));
```
The test failure is because this is now throwing `InvalidEnumArgumentException`. I can't see why this would be special and use `ArgumentOutOfRangeException`, its the odd one out in doing so.

Anyone know if this is a problem to change?

hughbe replied:
I kinda agree with changing this to `InvalidEnumArgumentException` but this would be breaking and has to be documented

RussKie replied:
> maybe we can delete all of this code in future: [dotnet/csharplang#3179](https://github.com/dotnet/csharplang/issues/3179)

This will probably another 3-4 years before we get this for GA.

davidwengier commented:
Maybe I should make this a nuget package and make my fortune in open source. If its good enough for NerdBank...

davidwengier commented:
I'm marking this as ready for review. It's basically done as far as replacing the previous validation.

Need to work out what to do with the two Asserts that are the only thing left using ClientUtils.IsEnumValid.

Probably could do with some more tests around generation too... Calling a validate method that isn't the enum validator, test the diagnostics reported.

davidwengier commented:
I don't know if its good news or bad news, but if we're lucky, maybe we can delete all of this code in future: https://github.com/dotnet/csharplang/issues/3179

RussKie commented:
> maybe we can delete all of this code in future: [dotnet/csharplang#3179](https://github.com/dotnet/csharplang/issues/3179)

This will probably be another 3-4 years before we get this for GA.

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/CheckBoxTests.cs:
```diff
@@ -290,38 +270,56 @@ public void CheckBox_OnClick_AutoCheckFalse_DoesNotChangeCheckState(bool threeSt
-
-            Assert.Equal(expected, box.CheckState);
+            Assert.Equal(value, control.CheckState);
+            // TODO: ControlAccessibleObject shouldn't force handle creation, tracked in https://github.com/dotnet/winforms/issues/3062
+            //Assert.False(control.IsHandleCreated);
+
+            // Set same.
+            control.CheckState = value;
+            Assert.Equal(value, control.CheckState);
+            // TODO: ControlAccessibleObject shouldn't force handle creation, tracked in https://github.com/dotnet/winforms/issues/3062
+            //Assert.False(control.IsHandleCreated);
```
Where did these come from? I can't see them in the original code.
Nor we dealing here with accessibility objects, so these comments here are very surprising.

davidwengier replied:
@hughbe This is from the commit you asked me to cherry pick which had the better invalid enum theory data generation.

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ControlTests.Methods.cs:
```diff
@@ -2099,38 +2099,26 @@ public void Control_GetAutoSizeMode_Invoke_ReturnsExpected()
-            Assert.Null(ret);
+            // Call again.
+            Assert.Null(control.GetChildAtPoint(new Point(5, 5), skipValue));
+            Assert.True(control.IsHandleCreated);
```
The 2nd check is redundant, once a handle is created it doesn't go away.
Checking twice that handle isn't created, on the other hand, is useful :)

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/TableLayoutSettings.cs:
```diff
@@ -73,10 +73,7 @@ internal TableLayoutPanelCellBorderStyle CellBorderStyle
             set
             {
                 //valid values are 0x0 to 0x6
-                if (!ClientUtils.IsEnumValid(value, (int)value, (int)TableLayoutPanelCellBorderStyle.None, (int)TableLayoutPanelCellBorderStyle.OutsetPartial))
-                {
-                    throw new ArgumentOutOfRangeException(nameof(value), value, string.Format(SR.InvalidArgument, nameof(CellBorderStyle), value));
-                }
```
❗ [@RussKie] to document

RussKie commented on src/System.Windows.Forms/src/System/Windows/Forms/TableLayoutSettings.cs:
```diff
@@ -206,10 +203,7 @@ public TableLayoutPanelGrowStyle GrowStyle
                 //valid values are 0x0 to 0x2
-                if (!ClientUtils.IsEnumValid(value, (int)value, (int)TableLayoutPanelGrowStyle.FixedSize, (int)TableLayoutPanelGrowStyle.AddColumns))
-                {
-                    throw new ArgumentOutOfRangeException(nameof(value), value, string.Format(SR.InvalidArgument, nameof(GrowStyle), value));
```
❗ [@RussKie] to document

RussKie reviewed (changes requested):
This is awesome 🚀 
Few questions and need few changes.

RussKie reviewed (approved):
This is awesome!
![image](https://user-images.githubusercontent.com/4403806/96834776-517d2680-148e-11eb-8b08-cc141cd9d85f.png)

RussKie commented:
Docs: https://github.com/dotnet/docs/issues/21239
